python/lsst/ip/isr/ampOffset.py

```diff
@@ -18,88 +18,406 @@
+                ampIm -= pedestal
+                metadata.set(
+                    f"PEDESTAL{ii + 1}", float(pedestal), f"Pedestal level subtracted from amp {ii + 1}"
+                )
+            self.log.info(f"amp pedestal values: {', '.join([f'{x:.4f}' for x in pedestals])}")
+
+        # raise NotImplementedError("Amp offset task should be retargeted
+        # by a camera specific version.")
+
+    def getAmpAssociations(self, amps):
+        """Determine amp geometry and amp associations from a list of
+        amplifiers.
+
+        Parse an input list of amplifiers to determine the layout of amps
+        within a detector, and identify all amp interfaces (i.e., the
+        horizontal and vertical junctions between amps).
+
+        Returns a matrix with a shape corresponding to the geometry of the amps
+        in the detector.
+
+        Parameters
+        ----------
+        amps : `list` [`lsst.afw.cameraGeom.Amplifier`]
+            List of amplifier objects used to deduce associations.
+
+        Returns
+        -------
+        ampAssociations : `numpy.ndarray`
+            A N x N matrix (N = the number of amplifiers) providing the
+            association information for amplifiers, indicating their
+            connections within the detector layout. The elements are defined as
+            below: TODO
+            -1: ...
+            ...: ...
+        ampInterfaces : `numpy.ndarray`
+            A N x N matrix (N = the number of amplifiers) representing the amp
+            interface information corresponding to the `ampAssociations`
+            matrix. The elements are integers defined as below:
+            -1: no interface
+            0: interface on the bottom
+            1: interfece on the right
+            2: interface on the top
+            3: interface on the left
+        """
+        xCenters = [amp.getBBox().getCenterX() for amp in amps]
+        yCenters = [amp.getBBox().getCenterY() for amp in amps]
+        xIndices = np.ceil(xCenters / np.min(xCenters) / 2).astype(int) - 1
+        yIndices = np.ceil(yCenters / np.min(yCenters) / 2).astype(int) - 1
+
+        nAmps = len(amps)
+        ampIds = np.zeros((len(set(yIndices)), len(set(xIndices))), dtype=int)
+
+        for ampId, xIndex, yIndex in zip(np.arange(nAmps), xIndices, yIndices):
+            ampIds[yIndex, xIndex] = ampId
+
+        ampAssociations = np.zeros((nAmps, nAmps), dtype=int)
+        ampInterfaces = np.full_like(ampAssociations, -1)
+
+        for ampId in ampIds.ravel():
+            adjacents, interfaces = self.getAdjacents(ampIds, ampId)
+            ampAssociations[ampId, adjacents] = -1
+            ampInterfaces[ampId, adjacents] = interfaces
+            ampAssociations[ampId, ampId] = -ampAssociations[ampId].sum()
+
+        if ampAssociations.sum() != 0:
+            raise RuntimeError("The `ampAssociations` array does not sum to zero.")
+
+        return ampAssociations, ampInterfaces
```

Comment:
Before this return, I think it would be nice to add the option for displaying these two matrices in a debug logger:
```
self.log.debug(f"amp associations:\n{ampAssociations}")
self.log.debug(f"amp interfaces:\n{ampInterfaces}")
```
which produces an output like this:
```
DEBUG 2023-08-01T12:10:55.114-07:00 lsst.isr.ampOffset (isr:{instrument: 'HSC', detector: 42, exposure: 1228, ...})(ampOffset.py:260) - amp associations:
[[ 1 -1  0  0]
 [-1  2 -1  0]
 [ 0 -1  2 -1]
 [ 0  0 -1  1]]
DEBUG 2023-08-01T12:10:55.117-07:00 lsst.isr.ampOffset (isr:{instrument: 'HSC', detector: 42, exposure: 1228, ...})(ampOffset.py:261) - amp interfaces:
[[-1  1 -1 -1]
 [ 3 -1  1 -1]
 [-1  3 -1  1]
 [-1 -1  3 -1]]
```

Reply:
Actually, we should probably move away from using f-strings in loggers, as it causes the log messages to be called even when not shown. The alternative which is better here (and every other log message in this file) should look like this instead:
```
self.log.debug("amp associations:\n%s", ampAssociations)
self.log.debug("amp interfaces:\n%s", ampInterfaces)
```